src/protocols/ab/session.c

```diff
@@ -1408,28 +1410,42 @@ int process_requests(ab_session_p session)
             /* if there are still requests after purging all the aborted requests, process them. */
 
             /* how much space do we have to work with. */
-            remaining_space = session->max_payload_size - (int)sizeof(cip_multi_req_header);
+            remaining_request_space = session->max_payload_size - (int)sizeof(cip_multi_req_header);
+
+             /* -2 bytes for the msp (multi service packet) number of packets and -4 bytes for the cip response header,
+              * we later subtract 2 bytes for each packet, this is to allow for the INT offset to the packet within the msp */
+            remaining_response_space = session->max_payload_size - 2 - 4;
 
             if(vector_length(session->requests)) {
                 do {
                     request = vector_get(session->requests, 0);
 
-                    remaining_space = remaining_space - get_payload_size(request);
+                    remaining_request_space = remaining_request_space - get_payload_size(request);
+
+                    /* calculate if all of the response data from the packed requests will fit into a single response packet. 
+                     * the -8 bytes is the maximum padding between packets, the -2 bytes is from the offset integer which stores the
+                     * offset to this packet */
+                    remaining_response_space = remaining_response_space - request->response_capacity - 8 - 2; 
+
+                    /* The tag must have packing enabled and the plc must either support fragmented reads or this tag must have been read 
+                     * before, so that its response size is known */
+                    allow_packing = request->allow_packing && (request->supports_fragmented_read || !request->first_read);
```

Comment:
I am not sure why you are doing this logic.  Is this to make sure we know the size of the tag first?   Hmm, that makes sense.   Let me know if this is the right guess.

Reply:
That's correct, this is to make sure that we either know the size of the response data we are requesting (so we know it will fit in the buffer), or that the PLC supports fragmented reads in which case it doesn't matter.